### PR TITLE
pss: Refactor. Step 1. Refactor PssMsg

### PR DESCRIPTION
**Pull request by jpeletier**
As part of the PSS module refactor effort, This PR moves `PssMsg` and `Topic` to its own package for clarity, modularity and separation of concerns.

Changes include:

- Refactored `msgParams` as `message.Flags` with its own `rlp.Decoder/Encoder` interface implementation.
- Removed `Serialize` method as it was only being used to compute a `Digest` cache key. Created a `Digest` method that computes a message digest efficiently.
- Removed `isSym()` and `isRaw()` methods, since the `Flags` struct is accesible directly.
- Renamed some methods for readability.
- Removed unused code.
- Wrote unit tests (92% coverage), with RLP encoding/decoding and JSON marshalling tests, etc, by means of Epic's `ut` toolkit.

### CODE REVIEW

File: pss/crypto/crypto.go
Comment (nolash) [inline pss/crypto/crypto.go]: This is just a moved file? No changes?

File: pss/forwarding_test.go
Comment (nolash) [inline pss/forwarding_test.go]: If we have other `utils` than crypto later maybe we should leave name `crypto` in this symbol for clarity

File: pss/keystore.go
Comment (nolash) [inline pss/keystore.go]: Why do we need a getter, please?

File: pss/message/message.go
Comment (nolash) [inline pss/message/message.go:32]: Please consider adding expire to the digest. That will allow intentional resends of messages.

File: pss/message/message.go
Comment (nolash) [inline pss/message/message.go:45]: Please consider adding expire

File: pss/message/message_test.go
Comment (nolash) [inline pss/message/message_test.go]: Please consider sourcing address from already existing methods that generate addresses, for example `network.RandomAddress()` (which in turn could use a deterministic seed as you use here in `RandomArray`)

Or even better would be if we had an address constructor in `pss` itself, since `pss` also operates with partial addressing.

I would prefer if the different addresses we use in Swarm are _strongly_ typed.

File: pss/message/message_test.go
Comment (nolash) [inline pss/message/message_test.go]: How did you generate the results for this, please?

File: pss/pss_test.go
Comment (nolash) [inline pss/pss_test.go:85]: This is an ugly relic, indeed. Maybe we can just `fmt.Sprintf` this? I don't think `strings.Join` and `strconv` is used like this in _any_ other test in swarm.

File: pss/pss_test.go
Comment (nolash) [inline pss/pss_test.go:146]: These error messages suggest something is _stored_, but that's not the case any longer, I believe.

**Review by nolash — CHANGES_REQUESTED** [review]

File: pss/crypto/crypto.go
Comment (jpeletier) [reply pss/crypto/crypto.go]: Sorry, these changes belong to #1731. This PR shouldn't be reviewed until that one is merged.

File: pss/forwarding_test.go
Comment (jpeletier) [reply pss/forwarding_test.go]: Sorry, these changes belong to #1731. This PR shouldn't be reviewed until that one is merged.

File: pss/keystore.go
Comment (jpeletier) [reply pss/keystore.go]: Sorry, these changes belong to #1731. This PR shouldn't be reviewed until that one is merged.

File: pss/message/message.go
Comment (jpeletier) [reply pss/message/message.go:32]: Not part of this PR. This PR is a refactor and does not add functionality. `Expire` was not part of the digest.

File: pss/message/message.go
Comment (jpeletier) [reply pss/message/message.go:45]: Not part of this PR. This PR is a refactor and does not add functionality. Expire was not part of the digest.

File: pss/pss_test.go
Comment (jpeletier) [reply pss/pss_test.go:85]: I am not refactoring these tests, just carving `PssMessage` out.

File: pss/pss_test.go
Comment (jpeletier) [reply pss/pss_test.go:146]: Again, not part of the changes in this PR.

File: pss/keystore.go
Comment (kortatu) [reply pss/keystore.go]: ReceivedMessage must be an interface because the validation and parsing is implementation-specific.

File: pss/message/message_test.go
Comment (jpeletier) [reply pss/message/message_test.go]: `pot.RandomAddress()` does not currently generate addresses based on a seed, and I need that for deterministic testing, hence the use of `RandomArray()` for now.

Opened #1735 to track a change to that method so it can be used.

> Or even better would be if we had an address constructor in pss itself, since pss also operates with partial addressing.

We can visit this when visiting `PssAddress` during the refactor.

> I would prefer if the different addresses we use in Swarm are strongly typed.

Me too, but I try to keep the scope of the refactor to `PssMsg` exclusively, otherwise the PR scope creeps and then it becomes a big PR.

File: pss/message/message_test.go
Comment (jpeletier) [reply pss/message/message_test.go]: At the begining of the test method, there is a flag you can set to `true` which will generate all test results if you run the test. When you do that, the test will be marked as failed, but it will have generated all values. You can then revert the flag back to `false` and the stored results are verified against.

You would only do that if after changing the `Message` structure for some reason, e.g., a protocol upgrade, and want the tests to be updated automatically. This avoids having strings all over the code and having to manually edit them when there is the slightest change (e.g., incorporating `Expires` to the digest, which will happen soon 😉).

Results tested with `EqualsKey()` are stored under `testdata/<Test Name>/results.json` under the corresponding key, in this case `msg0-rlp`, `msg1-rlp`... and so on. 

(`testdata` folders are special and skipped by Go tools)

File: pss/message/message.go
Comment (jpeletier) [reply pss/message/message.go:32]: Added #1736 to track this. Should be written and prioritized as a user story.

**Review by zelig — COMMENTED** [review]
LGTM, will submit review after #1731 merged

**Review by zelig — CHANGES_REQUESTED** [review]
can we not introduce unnecessary complexity with a testing framework please?
otherwise lgtm

File: pss/message/topic.go
Comment (nolash) [reply pss/message/topic.go:41]: It marshals to b64 if not

File: pss/message/message_test.go
Comment (nolash) [reply pss/message/message_test.go]: A question: Is there some risk associated with using the same code that is to be tested as the source of truth for the results it is supposed to give?

I am aware that the way we are doing it currently, eg. by copying the new hash from terminal output and pasting it into the test, probably doesn't provide the best guarantee either. I just can't help wondering that automation somehow invites the dingo.

File: pss/message/testdata/TestFlags/results.json
Comment (nolash) [reply pss/message/testdata/TestFlags/results.json]: I agree. I often find our test files rather cluttered.

File: pss/message/message_test.go
Comment (nolash) [reply pss/message/message_test.go]: That aside:

If I remember correctly, our conclusion from discussing this half a year ago was that @jpeletier should feel free to try out the tool confined to the feeds package, as this more or less was a package he alone was developing anyway.

Also, I think the controversy concerning the package was _not_ the test result data generation part of it, but rather that the assertion tools seemed superfluous. IIRC @nonsense was particularly outspoken there.

I am not against trying this out in pss instead, as @kortatu and @aperonam and he all are familiar with the framework and actively make use of it in their development work in general. But perhaps we for now should confine it to this package, at least before we have another team-wide discussion about introducing it.

**Review by nolash — APPROVED** [review]

File: pss/message/message_test.go
Comment (jpeletier) [reply pss/message/message_test.go]: > A question: Is there some risk associated with using the same code that is to be tested as the source of truth for the results it is supposed to give?

It is not exactly like this. The way to use it is: you run the test code for the first time, it produces a set of results **which is your responsibility to manually check** and sign off to make sure they make sense (just as you now would type the expected string or value, or calculate the expected hash somewhere else, which you then copy and paste in your code). Once you're happy with the values, you then "freeze" that behavior.

> [...] probably doesn't provide the best guarantee either. I just can't help wondering that automation somehow invites the dingo.

You are not forced to have automated generation, you can type up the results file by hand.

In my opinion the net result is: less time actually writing test code, more time to write for more coverage and test quality.

Comment (jpeletier) [reply pss/message/message_test.go]: > [...] the assertion tools seemed superfluous. [...]

That is a point of view. My point of view is that any overhead work involved in setting the stage for a condition to be tested a) Takes time to write b) is prone to errors in the test code itself, which takes also time to debug and c) Hides the purpose of the test in the "bureaucracy". All of these make the test harder to read and evolve over time.

> If I remember correctly, our conclusion from discussing this half a year ago was that @jpeletier should feel free to try out the tool confined to the feeds package, as this more or less was a package he alone was developing anyway.

Yes, I am moving that freedom to pss instead, this is not intended to extend generally outside of this package until discussed in depth.

Comment (zelig) [reply pss/message/message_test.go]: guys, I strongly resist this. My 2cents:

- it is generally against go philosophy - readability and focussed tests can be achieved by setup helpers etc,
- such (esp branded) dependencies are unnecessary risk factor due to maintanance/updates/continuity/compatibility
- any mediation in test code is extra risk, because errors can be deeply hidden in the framework (your point b) --- so testing frameworks should only be used in special cases where they genuinely bring needed functionality (network simulation)
- might have saved your team time but would create extra work for the current team  -- so time to learn/master (vs  your point a)
- i would like to see convergence of code/test style not more divergence, so i am not very keen  on introducing it in one place
- autogenerated or not, having few byte files to test a simple serialisation is no better than inline 
-  changing asset directory names if you change testname also not too good

I'd like to hear others opinions, @janos , @lmars @nonsense , so i add it to the roundtable topics for tuesday. https://hackmd.io/oj9_cT2KQimMdIPe_W_ejQ?both

Comment (lmars) [reply pss/message/message_test.go]: One of the main reasons I use Go is because of the simplicity and heavy focus in the community of doing things one way with little focus on customisation. For example, whilst I prefer double spacing to tabs, I use tabs, because that's what `gofmt` does, and everyone uses (or should use :wink:) `gofmt`.

Everyone doing things one way means I can jump into any codebase and immediately feel familiar with the coding style and tooling used and focus on the business logic. This is an important property for any sizeable software project to have where developer onboarding / churn becomes more of a concern over time and focus on business logic rather than bike-shedding is a must.

Applying this to the idea of "test helper" packages, I typically avoid them because the stdlib `testing` package is what most people use, and it gives me what I need to write effective tests.

From a cursory glance at the microTest README, I see things like:

```go
Ok(t, err)

Assert(t, x == y, "expected x to equal %v, got %v", y, x)

Equals(t, x, y)
```

which I would immediately argue should just be:

```go
if err != nil {
	t.Fatal(err)
}

if x != y {
	t.Fatalf("expected x to equal %v, got %v", y, x)
}

if x != y {
	t.Fatal("values aren't equal")
}
```

They may require a few more lines and, for those without editor macros, a few more key strokes, but I'd argue the time / space lost is far outweighed by the clarity of doing it the same way everyone else does it.

On this point:

> one important feature is the ability to generate and store test results for later verification

I personally dislike auto-fixture generation style tests, again because I feel it is simpler and more readable to just have a function at the start of a test that either declares or generates the fixture data. I see no issue with this, for example:

```go
func TestFoo(t *testing.T) {
	var expectedFoo = []byte(`<some fixture data>`)
	foo, err := Foo()
	if err != nil {
		t.Fatal(err)
	}
	if foo != expectedFoo {
		t.Fatalf("expected foo to be %v, got %v", expectedFoo, foo)
	}
}
```

This is my 2cents from just reading the above comment thread.

**Comment by nolash** [comment]
@jpeletier please fix linter error

File: pss/message/message_test.go
Comment (janos) [reply pss/message/message_test.go]: This is only my point of view about testing in Go. And I saw that the third party package was removed.

Using testing package should be the preferred way instead a generalized helper functions in a third party utility package. Simple checks like Ok, Assert, Equal do not bring so much clarity, as it adds an indirection. Helper functions should be written for the package that is tested for abstracting more complex validations. If the package need to provide some unified way of testing other implementations or its types or behaviour in other packages, then a dedicated testing package for this purpose should be made. The generalized testing package on the project level should be as small as possible without dependencies on packages in that project.

Comment (jpeletier) [reply pss/message/message_test.go]: I have removed these dependencies. If there are no other outstanding concerns, this can be closed.

**Review by zelig — APPROVED** [review]

**Pull request merged**
